cloud/shared/bin/lib/terraform.py

```diff
@@ -1,37 +1,95 @@
 import subprocess
 import os
+import re
 import shutil
 import shlex
+import inspect
 from typing import Optional
 
 from cloud.shared.bin.lib.config_loader import ConfigLoader
 from cloud.shared.bin.lib.print import print
 
 
-# TODO(#2741): When using this for Azure make sure to setup backend bucket prior to calling these functions.
-def perform_apply(
+def force_unlock(
         config_loader: ConfigLoader,
-        is_destroy=False,
+        lock_id: str,
         terraform_template_dir: Optional[str] = None):
-    '''Generates terraform variable files and runs terraform init and apply.'''
     if not terraform_template_dir:
         terraform_template_dir = config_loader.get_template_dir()
-    tf_vars_filename = config_loader.tfvars_filename
 
-    terraform_cmd = f'terraform -chdir={terraform_template_dir}'
+    perform_init(config_loader, terraform_template_dir, False)
+
+    terraform_cmd = f'terraform -chdir={terraform_template_dir} force-unlock -force {lock_id}'
+    print(f" - Run {terraform_cmd}")
+    subprocess.check_call(shlex.split(terraform_cmd))
+    return True
+
+
+def perform_init(
+        config_loader: ConfigLoader,
+        terraform_template_dir: Optional[str] = None,
+        upgrade: bool = True):
+    if not terraform_template_dir:
+        terraform_template_dir = config_loader.get_template_dir()
+
+    init_cmd = f'terraform -chdir={terraform_template_dir} init'
+    if upgrade:
+        init_cmd += ' -upgrade'
 
     if config_loader.use_local_backend:
-        print(' - Run terraform init -upgrade -reconfigure')
-        subprocess.check_call(
-            shlex.split(f'{terraform_cmd} init -upgrade -reconfigure'))
+        init_cmd += ' -reconfigure'
     else:
-        print(' - Run terraform init -upgrade')
-        init_cmd = f'{terraform_cmd} init -input=false -upgrade'
+        init_cmd += ' -input=false'
         # backend vars file can be absent when pre-terraform setup is running
         if os.path.exists(os.path.join(terraform_template_dir,
                                        config_loader.backend_vars_filename)):
             init_cmd += f' -backend-config={config_loader.backend_vars_filename}'
-        subprocess.check_call(shlex.split(init_cmd))
+    print(f" - Run {init_cmd}")
+    output, exit_code = capture_stderr(init_cmd)
+    if exit_code:
+        # This is AWS-specific, and should be modified when we have actual
+        # Azure deployments
+        if 'state data in S3 does not have the expected content' in output:
+            match = re.search(r'value: ([0-9a-f]{32})', output)
+            if match:
+                print(
+                    f"To fix the above error, rerun this command with \"--fix-digest={match.group(match.lastindex)}\""
+                )
+            exit(exit_code)
```

Comment:
**suggestion (blocking):** let's not fail silently for other cloud providers. Use `config_loader.get_cloud_provider()` to ensure the provider is AWS, throw `NotImplementedError` if it's not